File: Version0.3/Matlab/Preprocessing/mkv2t.m
function [re,ce] = mkv2t(t,ne)

sz = size(t);
if sz(1)==ne
    t = t';
end

ndof = max(t(:));
re = zeros(ndof,1); % store number of neighboring elements for each entity
for i = 1:ne
    k = t(:,i);
    k = k(k>0);
    re(k) = re(k) + 1;
end
re=[0; cumsum(re)];

ce = zeros(re(end),1);  % store neighboring-element indices for each entity
in = ones(ndof,1);
for i = 1:ne
    k = t(:,i);   % entities on element i  
    k = k(k>0);
    % re(k): pointer to the list of entities k
    ce(re(k)+in(k)) = i;
    in(k) = in(k) + 1;
end

File: Version0.1/Python/Preprocessing/mkv2t.py
from numpy import *
from getelemface import *

def  mkv2t(t,ne):

    ndof = max(t.flatten())+1;
    re = zeros(ndof).astype(int); # store number of neighboring elements for each entity
    for i in range(0,ne):
        k = t[:,i];
        k = k[k>-1];
        re[k] = re[k] + 1;

    re = hstack((array([0]), cumsum(re)));
    re = re.astype(int);

    ce = zeros(re[-1]).astype(int);  # store neighboring-element indices for each entity
    ind = ones(ndof).astype(int);
    for i in range(0,ne):
        k = t[:,i];   # entities on element i
        k = k[k>-1];
        # re(k): pointer to the list of entities k
        ce[re[k]+ind[k]-1] = i+1;
        ind[k] = ind[k] + 1;

    return re,ce
